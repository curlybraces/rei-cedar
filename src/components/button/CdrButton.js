export { default as CdrButton } from './CdrButton.jsx';
