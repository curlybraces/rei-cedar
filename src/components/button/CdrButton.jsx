import modifier from 'mixinsdir/modifier';
import size from 'mixinsdir/size';
import space from 'mixinsdir/space';
import cs from 'classnames';
import style from './styles/CdrButton.scss';

export default {
  name: 'CdrButton',
  mixins: [modifier, size, space],
  props: {
    /**
     * Controls render as button or anchor. {button, a}
     */
    tag: {
      type: String,
      default: 'button',
      validator: value => (['button', 'a'].indexOf(value) >= 0) || false,
    },
    /**
     * Sets the button type. {button, submit, reset}
     */
    type: {
      type: String,
      default: 'button',
      validator: value => (['button', 'submit', 'reset'].indexOf(value) >= 0) || false,
    },
    /**
     * Sets width to be 100%.
    */
    fullWidth: {
      type: Boolean,
      default: false,
      validator: value => typeof value === 'boolean',
    },
    /**
     * Renders an icon-only button. Default slot is disabled. Overrides size and responsiveSize props.
     */
    iconOnly: {
      type: Boolean,
      default: false,
    },
    /**
     * Renders an icon-only button with a light fill color for use on dark backgrounds.
     * iconOnly must be true.
     */
    onDark: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-button';
    },
    defaultClass() {
      return this.modifyClassName(this.baseClass, 'primary');
    },
    buttonSizeClass() {
      return !this.iconOnly ? this.sizeClass : null;
    },
    iconClass() {
      const classes = [];

      if (this.$slots.icon && this.$slots.default) {
        /* only add class for buttons with text + icon */
        classes.push(this.modifyClassName(this.baseClass, 'has-icon'));
      }

      if (this.iconOnly) {
        classes.push(this.modifyClassName(this.baseClass, 'icon-only'));

        if (this.onDark) {
          classes.push(this.modifyClassName(this.baseClass, 'on-dark'));
        }
      }
      return classes.join(' ');
    },
    fullWidthClass() {
      return this.fullWidth && !this.iconOnly
        ? this.modifyClassName(this.baseClass, 'full-width') : null;
    },
  },
  render() {
    const Component = this.tag;
    return (<Component
      class={cs(this.defaultClass,
        this.modifierClass,
        this.buttonSizeClass,
        this.fullWidthClass,
        this.iconClass,
        this.space)}
      type={this.tag === 'button' ? this.type : null}
      {...{ on: this.$listeners }}
    >
      {this.$slots.icon}
      {this.$slots.default}
    </Component>);
  },
};
