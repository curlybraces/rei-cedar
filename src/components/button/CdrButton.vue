<template>
  <component
    :is="tag"
    :class="[blockClass, sizeClass, fullWidthClass, responsiveClass, styleModifiersClass]"
    :type="tag === 'button' ? type : null"
    @click="onClick"
  >
    <!-- @slot innerHTML on the inside of the button component -->
    <slot/>
  </component>
</template>

<script>
import classModifier from 'mixinsdir/classModifier';
import buttonBase from 'mixinsdir/buttonBase';

/**
 * Cedar 2 component for button
 *
 * CdrButton will render either a button, or a anchor that looks like a button.
 * As such, the decision to use CdrButton vs CdrAnchor should be made based on what
 * you need the rendered element to look like.
 *
 * @version 0.1.0
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrButton',
  mixins: [buttonBase, classModifier],
  props: {
    /**
     * Sets a static size for the button, which scales padding and text size. {small, medium, large}
     */
    size: {
      type: String,
      default: 'medium',
      validator: value => (['small', 'medium', 'large'].indexOf(value) >= 0) || false,
    },
    /**
     * Sets width to be 100%. Can be combined with staticSize.
    */
    fullWidth: {
      type: Boolean,
      default: false,
      validator: value => typeof value === 'boolean',
    },
    /**
     * Render a specific button size at a specific breakpoint. Takes precedence over staticSize and fullWidth.
     * Default function returns an empty array.
     */
    responsiveSize: {
      type: Array,
    },
    /**
     * Additional style modifiers. Default function returns an empty array.
     * {`secondary`}
     */
    styleModifiers: {
      type: String,
      validator: value => value === 'secondary',
    },
  },
  computed: {
    blockClass() {
      return 'cdr-button';
    },
    sizeClass() {
      return this.modifierClass(this.size);
    },
    fullWidthClass() {
      return this.fullWidth ? this.modifierClass('full-width') : null;
    },
    responsiveClass() {
      const responsiveClass = [];

      if (this.responsiveSize) {
        this.responsiveSize.forEach((val) => {
          responsiveClass.push(this.modifierClass(val));
        });
      }

      return responsiveClass.join(' ');
    },
    styleModifiersClass() {
      return this.styleModifiers ? this.modifierClass(this.styleModifiers) : null;
    },
  },
};
</script>

<style>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/CdrButton.pcss';
</style>
