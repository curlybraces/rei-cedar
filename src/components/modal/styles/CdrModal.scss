@import '../../../css/settings/index.scss';

$duration: 150ms;
$smMin: 576px;
$overlayColor: rgba(0, 0, 0, .4);
$contentColor: #fff;
$modalBorderRadius: .5rem;

:export {
  animationDuration: 300;
}

.modal {
  bottom: 0;
  height: 100%;
  left: 0;
  overflow-y: scroll;
  position: fixed;
  right: 0;
  top: 0;
  visibility: visible;
  z-index: 1000;
}

.closed {
  transition: z-index 0s $duration * 2,
  visibility 0s $duration * 2;
  visibility: hidden;
  z-index: -1;

  .overlay {
    opacity: 0;
    transition: opacity $duration $duration;
  }

  .contentWrap {
    opacity: 0;
    transition: opacity $duration;
  }
}


.overlay {
  background-color: $overlayColor;
  bottom: 0;
  left: 0;
  opacity: 1;
  position: fixed;
  right: 0;
  top: 0;
  transition: opacity $duration;
  z-index: -1;
}

.outerWrap {
  display: flex;
  box-sizing: border-box;
  left: 0;
  min-height: 100%;
  padding: 16px 20px;
  position: absolute;
  right: 0;
  top: 0;
  z-index: -1;
}

.noscroll {
  overflow: hidden !important;
  position: fixed !important;
  width: 100%;
}

.contentWrap {
  align-items: flex-start;
  background-color: $contentColor;
  border-radius: $modalBorderRadius;
  box-shadow: 0 .5rem 1.5rem rgba(0, 0, 0, .5);
  display: flex;
  flex-direction: column;
  margin: auto;
  max-width: 40rem;
  opacity: 1;
  padding: 20px;
  position: relative;
  transition: opacity $duration $duration;
  width: 100%; // IE 11
  z-index: 0;
  &:focus {
    outline: 0;
  }
}

.innerWrap {
  width: 100%; // IE 11
}

@media (min-width: $smMin) {
  .content-wrap {
    padding: 1.6rem;
  }
}

$dialog-margins-vertical: $cdr-space-four-x * 2; // 12.8rem
$dialog-padding-vertical: $cdr-space-one-and-a-half-x; // 2.4rem
$header-padding-bottom: $cdr-space-one-x; // 1.6rem
$title-line-height: $cdr-space-two-x; // 3.2rem

// total up internal padding to get max height of text
$header-height: $header-padding-bottom + $title-line-height;
$modal-total-padding: #{$dialog-margins-vertical + $header-height + $dialog-padding-vertical * 2};
$modal-text-max-height: calc(100vh - #{$modal-total-padding});

.pdp-modal {
  &__dialog {
    max-height: calc(100vh - #{$dialog-margins-vertical});
    max-width: 60rem;
    padding: $dialog-padding-vertical;

     &:global(.large) {
      max-width: 80rem;
    }
  }

  &__content {
    position: relative;
  }

  &__header {
    display: flex;
    padding: 0 0 $header-padding-bottom 0;
  }

  &__title {
    flex: auto;
    padding-right: $cdr-space-three-quarter-x;
  }

  &__close-button {
    flex: none;
    padding: 0 0 0 $cdr-space-quarter-x !important;
    margin-right: -$cdr-space-half-x;
  }

  &__text {
    padding: 0;
    position: relative;
  }

  &__text-content {
    max-height: $modal-text-max-height;
    overflow: auto;

    /* ie/edge/ff bug prevents using padding */
    &::after {
      content: "";
      display: block;
      height: $cdr-space-two-x;
      width: 100%;
    }
  }

  &__text-fade {
    position: absolute;
    bottom: 0;
    left: 0;
    height: $cdr-space-two-x;
    width: 98%;
    /* full transparency doesn't fly on safari */
    background: linear-gradient(rgba(255,255,255,0.001), rgba(255,255,255,1));
    background-attachment: scroll;
  }
}
