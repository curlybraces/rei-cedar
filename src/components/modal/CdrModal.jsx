import tabbable from 'tabbable';
import style from './styles/CdrModal.scss';
import onTransitionEnd from './onTransitionEnd';
import CdrButton from '../button/CdrButton';
import CdrIcon from '../icon/CdrIcon';
import CdrText from '../text/CdrText';
import clsx from 'clsx';

export default {
  name: 'CdrModal',
  components: {
    CdrButton,
    CdrIcon,
    CdrText,
  },
  props: {
    opened: {
      type: Boolean,
      required: true,
    },
    closeModal: {
      type: Function,
      required: true,
    },
    label: {
      type: String,
      required: true,
    },
    labelIsTitle: {
      type: Boolean,
      required: false,
      default: false,
    },
    ariaDescribedbBy: {
      type: String,
      required: false,
      default: null,
    },
    id: {
      type: String,
      required: false,
      default: null,
    },
    overlayClass: String,
    wrapperClass: String,
    contentClass: String,
    animationDuration: {
      type: Number,
      default: parseInt(style.animationDuration, 10),
    },
    width: {
      type: String,
      required: false,
      default: 'medium',
      validator: val => val === 'medium' || val === 'large',
    },
  },
  data() {
    return {
      style,
      unsubscribe: null,
      keyHandler: null,
      lastActive: null,
      focusHandler: null,
      reallyClosed: !this.opened,
      offset: null,
    };
  },
  computed: {
    dialogAttrs() {
      return {
        'aria-describedby': this.ariaDescribedBy,
        'aria-modal': 'true',
        id: this.id,
      };
    },
    dialogClass() {
      return `pdp-modal__dialog ${this.width}`;
    },
  },
  watch: {
    opened(newValue, oldValue) {
      if (!!newValue === !!oldValue) return;
      if (newValue) {
        this.handleOpened();
      } else {
        this.handleClosed();
      }
    },
  },
  mounted() {
    if (this.opened) {
      this.addNoScroll();
      this.addHandlers();
    }
  },
  beforeDestroy() {
    if (this.unsubscribe) this.unsubscribe();
    if (this.opened) this.removeNoScroll();
    document.removeEventListener('focusin', this.focusHandler, true);
    document.removeEventListener('keydown', this.keyHandler);
  },
  methods: {
    handleKeyDown({ key }) {
      switch (key) {
        case 'Escape':
          this.closeModal();
          break;
        default: break;
      }
    },
    handleFocus(e) {
      console.log('handleFocus', e);
      const { documentElement } = document;
      if (this.$refs.modal.contains(e.target) || !documentElement) return;

      const tabbables = tabbable(documentElement);
      const these = tabbable(this.$refs.modal);
      const nextIx = tabbables.indexOf(e.target);
      const firstModalIx = tabbables.indexOf(these[0]);
      const nextRef = nextIx < firstModalIx ? these[these.length - 1] : these[0];
      if (nextRef) nextRef.focus();

      documentElement.scrollTop = this.scrollTop;
      documentElement.scrollLeft = this.scrollLeft;
    },
    handleOpened() {
      // console.log('handleOpened');
      const { activeElement } = document;
      // console.log('activeelement', activeElement);

      this.addNoScroll();
      this.reallyClosed = false;
      this.lastActive = activeElement;

      this.$nextTick(() => {
        this.$refs.modal.focus();
        this.addHandlers();

        setTimeout(() => {
          // for some reason Safari scrolls the wrapper down a bit?
          // doesn't work without setTimeout for some unknown reason
          this.$refs.wrapper.scrollTop = 0;
        });
      });
    },
    handleClosed() {
      document.removeEventListener('keydown', this.keyHandler);

      this.unsubscribe = onTransitionEnd(
        this.$refs.wrapper,
        () => {
          console.log('ontTransitionEnd');
          this.unsubscribe();
          this.removeNoScroll();
          this.unsubscribe = null;
          this.reallyClosed = true;

          // restore previous scroll position
          window.scrollTo(this.offset.x, this.offset.y);

          document.removeEventListener('focusin', this.focusHandler, true);

          if (this.lastActive) this.lastActive.focus();
          // console.log('ontransitionEnd');
        },
        this.animationDuration + 16,
      );
    },
    addNoScroll() {
      const { documentElement, body } = document;
      const offset = {
        x: window.scrollX
          || (documentElement || {}).scrollLeft
          || (body || {}).scrollLeft
          || 0,
        y: window.scrollY
          || (documentElement || {}).scrollTop
          || (body || {}).scrollTop
          || 0,
      };
      this.offset = offset;

      if (documentElement) {
        documentElement.classList.add(style.noscroll);
        // keep current scroll position manually
        documentElement.style.top = `-${offset.y}px`;
        documentElement.style.left = `-${offset.x}px`;
      }

      if (body) {
        body.classList.add(style.noscroll);
      }
    },
    removeNoScroll() {
      const { documentElement, body } = document;

      if (body) {
        body.classList.remove(style.noscroll);
      }

      if (documentElement) {
        documentElement.classList.remove(style.noscroll);
        documentElement.style.top = '';
        documentElement.style.left = '';
      }
    },
    addHandlers() {
      this.focusHandler = this.handleFocus.bind(this);
      this.keyHandler = this.handleKeyDown.bind(this);
      document.addEventListener('focusin', this.focusHandler, true);
      document.addEventListener('keydown', this.keyHandler);
    },
  },
  render() {
    const {
      closeModal,
      modalId,
      opened,
      label,
      wrapperClass,
      overlayClass,
      dialogClass,
      contentClass,
      reallyClosed,
      $slots: {
        default: defaultSlot,
      },
    } = this;
    return (
      <div
        class={{
          [style.modal]: true,
          [style.closed]: !opened,
        }}
        ref="wrapper"
        role="presentation"
      >
        <div class={[style.outerWrap, wrapperClass]}>
          <div
            aria-hidden="true"
            onClick={closeModal}
            class={[style.overlay, overlayClass]}
          />
          {/* eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex */}
          <div tabIndex={opened ? '0' : undefined} />
          {/*
            this is the "uncommon case of an element that captures incoming tab traversal
            for a composite widget" mentioned in the rule docs
          */}
          <div
            ref="modal"
            class={[style.contentWrap, dialogClass]}
            id={modalId}
            tabIndex="-1"
            role="dialog"
            aria-modal={!!opened}
            aria-label={label}
            {...{ attrs: this.dialogAttrs || {} }}
          >
            <div
              class={[style.innerWrap, contentClass]}
              style={reallyClosed
                ? { display: 'none' }
                : undefined
              }
            >
              <section>
                <div class={this.style['pdp-modal__content']}>
                  <div class={this.style['pdp-modal__header']}>
                    <div class={this.style['pdp-modal__title']}>
                      {this.labelIsTitle && (
                        <cdr-text
                          tag="h1"
                          modifier="heading-600"
                        >
                          {this.label}
                        </cdr-text>
                      )}
                    </div>
                    <cdr-button
                      id="close-modal-button"
                      class={this.style['pdp-modal__close-button']}
                      icon-only={true}
                      on-click={closeModal}
                      aria-label="close"
                    >
                      <cdr-icon
                        slot="icon"
                        class={this.style['cdr-button__icon']}
                        use="#x-lg"
                        inherit-color
                      />
                    </cdr-button>
                  </div>
                  <div
                    role="document"
                    tabindex="0"
                    class={this.style['pdp-modal__text']}
                  >
                    <div class={this.style['pdp-modal__text-content']}>
                      {defaultSlot}
                    </div>
                    <div class={this.style['pdp-modal__text-fade']} />
                  </div>
                </div>
              </section>
            </div>
          </div>
          {/* eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex */}
          <div tabIndex={opened ? '0' : undefined} />
        </div>
      </div>
    );
  },
};
