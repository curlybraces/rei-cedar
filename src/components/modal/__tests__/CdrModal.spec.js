import { shallowMount, mount } from '@vue/test-utils';
import CdrModal from 'componentdir/modal/CdrModal';
import Vue from 'vue';
// import packageJson from '../../../../package.json';
import CdrButton from 'componentdir/button/CdrButton';

describe('CdrModal.vue', () => {
  it('renders all slots, handleClosed', () => {
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        title: 'Terms and Conditions',
        stickyContentSlot: 'Sticky content',
        scrollingContentSlot: 'Main content',
        footer: 'Footer content'
      },
      attachToDocument: true,
    });
    /*
      - begins opened
      - renders all slots for screenshots
      - close modal
    */

   expect(wrapper.element).toMatchSnapshot();
    Vue.nextTick(() => {
      wrapper.setProps({ opened: false });

      setTimeout(() => {
        expect(wrapper.vm.reallyClosed).toBe(true);
        wrapper.destroy();
      }, 300);
    });
  });

  it('leaves optional slots empty, handleOpened', () => {
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: false,
        label: "Label is the modal title"
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      attachToDocument: true,
    });

    wrapper.setProps({ opened: true });
    Vue.nextTick(() => {
      expect(wrapper.element).toMatchSnapshot();
      wrapper.destroy();
    });
  });

  it('handleKeyDown', () => {
    const spyOnClick = jest.fn();
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      methods: {
        onClick: spyOnClick
      },
      attachToDocument: true,
    });

    wrapper.trigger('keydown', {
      key: 'a'
    });
    expect(spyOnClick).not.toHaveBeenCalled();

    wrapper.trigger('keydown', {
      key: 'Esc',
    });

    wrapper.trigger('keydown', {
      key: 'Escape',
    });

    expect(spyOnClick).toHaveBeenCalledTimes(2);
    wrapper.destroy();
  });

  it('scrolling and fullscreen snapshot', () => {
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "Label is the modal title"
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      attachToDocument: true,
    });

    Vue.nextTick(() => {
      wrapper.setProps({ fullscreen: true, scrollHeight: 500, offsetHeight: 400 });
      setTimeout(() => {
        expect(wrapper.vm.scrolling).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
        wrapper.destroy();
      }, 300);
    })
  });

  it('removeNoScroll', () => {
    const wrapper = shallowMount(CdrModal, {
      propsData: {
        opened: true,
        label: "My Modal Label",
      },
      slots: {
        scrollingContentSlot: 'Main content',
      },
      attachToDocument: true,
    });
    const { documentElement, body } = document;
    wrapper.vm.removeNoScroll();

    expect(documentElement.classList.contains('noscroll')).toBeFalsy();
    expect(body.classList.contains('noscroll')).toBeFalsy();

    wrapper.destroy();
  });

  it('handleFocus', () => {
    const spyMeasureContent = jest.fn();
    const wrapper = mount(CdrModal, {
      propsData: {
        opened: true,
        label: "My Modal Label",
      },
      attachToDocument: true,
      methods: {
        measureContent: spyMeasureContent,
      },
    });

    const button = wrapper.find('button').element;
    button.focus();
    expect(button).toBe(document.activeElement);

    wrapper.vm.handleFocus({ target: document.createElement('a') });
    expect(document.scrollTop).toBe(undefined);
    expect(document.scrollLeft).toBe(undefined);
    wrapper.destroy();
  });

  // it('handleClosed', (done) => {
  //   spyOn(document, 'removeEventListener');
  //   global.scrollTo = jest.fn();
  //   const spyMeasureContent = jest.fn();
  //   const spyHandleOpened = jest.fn();
    
  //   const wrapper = shallowMount(CdrModal, {
  //     propsData: {
  //       opened: true, 
  //       label: "My Modal Label",
  //     },
  //     methods: {
  //       measureContent: spyMeasureContent,
  //       handleOpened: spyHandleOpened,
  //     },
  //     attachToDocument: true,
  //   });

    

  //   expect(document.removeEventListener).toHaveBeenCalledWith('keydown', expect.anything());
    
  //   Vue.nextTick(() => {
  //     setTimeout(() => {
  //       expect(wrapper.vm.reallyClosed).toBe(true);
  //       expect(wrapper.vm.unsubscribe).toBe(null);
  //       expect(document.removeEventListener).nthCalledWith(2, 'focusin', expect.anything(), true);
  //       done();
  //     }, 1000);
  //   });
  // });

});