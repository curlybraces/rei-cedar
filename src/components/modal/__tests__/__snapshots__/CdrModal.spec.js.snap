// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrModal.vue renders correctly 1`] = `
<div
  class="cdr-modal"
  role="presentation"
>
  <div
    class="cdr-modal__outerWrap"
  >
    <div
      aria-hidden="true"
      class="cdr-modal__overlay"
    />
    <div
      tabindex="0"
    />
    <div
      aria-label="Label is the modal title"
      aria-modal="true"
      class="cdr-modal__contentWrap cdr-modal__dialog medium"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="cdr-modal__innerWrap"
      >
        <section>
          <div
            class="cdr-modal__content"
          >
            <div
              class="cdr-modal__header"
            >
              <div
                class="cdr-modal__title"
              >
                <cdr-text-stub
                  modifier="heading-serif-600"
                  space=""
                  tag="h1"
                >
                  Label is the modal title
                </cdr-text-stub>
              </div>
              <cdr-button-stub
                aria-label="close"
                class="cdr-modal__close-button"
                icononly="true"
                id="close-modal-button"
                size="medium"
                space=""
                tag="button"
                type="button"
              >
                <iconxlg-stub
                  inheritcolor="true"
                  size="medium"
                  space=""
                />
              </cdr-button-stub>
            </div>
            <div
              class="cdr-modal__text"
              role="document"
              tabindex="0"
            >
              <div
                class="cdr-modal__text-content"
              >
                Modal text content
              </div>
              <div
                class="cdr-modal__text-fade"
              />
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      tabindex="0"
    />
  </div>
</div>
`;
